### ISSUE TITLE
Inconsistent return type for getStoreId()

### ISSUE THREAD

**Issue by sreichel** [magento/magento2#18079]
### Summary

- Magento 2.*
- PHP7.1
- using strict type decalaration

`getStoreId()` returns a `string` for products, but `int` for categories that causes a fatal error in extensions code.

### Examples

Simplified code:

ObserverA:

https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Catalog/Model/Product.php#L521-L532

```
$product = $observer->getProduct();
$storeId = $product->getStoreId();
$result = $this->someMethod($storeId);
```
ObserverB:

https://github.com/magento/magento2/blob/2.3-develop/app/code/Magento/Catalog/Model/Category.php#L560-L573

```
$category = $observer->getCategory();
$storeId = $category->getStoreId();
$result = $this->someMethod($storeId);
```
Shared method

```
public funtion someMethod(int $storeId)
```

If i pass products store ID to `someMethod` i get a fatal error.

### Proposed solution

1. cast return type to `int` in product model
2. check all (80) `getStoreId()` methods to return always int value

Would create PR for "1." since I#ve not enough time to check all others.

Please advice.

**Comment by magento-engcom-team** [comment]
Hi @sreichel. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me $VERSION instance` 

where `$VERSION` is [version tags](https://github.com/magento/magento2/tags) (starting from 2.2.0+) or develop branches (for example: 2.3-develop). 
For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@sreichel do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [x] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [x] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `G1 Passed` will be added to the issue automatically. Please, edit issue description if needed, until label `G1 Passed` appears.</details>

- [x] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `G2 Passed` label to the issue by yourself.

- [x] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [x] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [x] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

- [x] 6. Add label `acknowledged` once verification is complete. 

- [x] 7. Make sure that automatic system confirms that report is acknowledged.

**Comment by magento-engcom-team** [comment]
@engcom-backlog-nazar Thank you for verifying the issue. Based on the provided information internal tickets `MAGETWO-95051`, `MAGETWO-95052` were created

**Comment by ghost** [comment]
Hi @sreichel , thank you for your report.
We've acknowledge the issue and added to our backlog.

**Comment by magento-engcom-team** [comment]
Hi @sreichel. Thank you for your report. 
The issue has been fixed in magento/magento2#18086 by @sreichel in 2.2-develop branch
Related commit(s):
 - [9d8b04a0c9700e8a5bfe65911c8151672f202de6](https://github.com/magento/magento2/commit/9d8b04a0c9700e8a5bfe65911c8151672f202de6)

The fix will be available with the upcoming 2.2.8 release.

**Comment by slavvka** [comment]
Hi @sreichel. Thank you for your report. 
The issue has been fixed in magento/magento2#18303 by @sreichel in 2.3-develop branch
Related commit(s):
 - [5411c0d3310b20888859f1e8cb83a1e78b54877e](https://github.com/magento/magento2/commit/5411c0d3310b20888859f1e8cb83a1e78b54877e)

The fix will be available with the upcoming 2.3.1 release.
